apps/prairielearn/src/news_items/index.js

```diff
@@ -1,65 +1,96 @@
-const jsonLoad = require('../lib/json-load');
-const namedLocks = require('@prairielearn/named-locks');
+import { logger } from '@prairielearn/logger';
+import * as namedLocks from '@prairielearn/named-locks';
+import * as sqldb from '@prairielearn/postgres';
+import * as Sentry from '@prairielearn/sentry';
+
+import * as schemas from '../schemas';
+import * as jsonLoad from '../lib/json-load';
+
+const DIRECTORY_REGEX = /^([0-9]+)_.+$/;
 
 async function loadNewsItems() {
   const news_items = [];
-  const regex = /^([0-9]+)_.+$/;
-  const dirs = await fsPromises.readdir(__dirname);
-  await async.each(dirs, async (dir) => {
-    const infoFilename = path.join(__dirname, dir, 'info.json');
-    let info;
-    try {
-      info = await jsonLoad.readInfoJSON(infoFilename, schemas.infoNewsItem);
-    } catch (err) {
-      if (err.code === 'ENOTDIR' || err.code === 'ENOENT') return; // skip dir entries without an info.json
-      throw err;
-    }
-    info.directory = dir;
-    info.index = regex.exec(dir)[1];
-    news_items.push(info);
-  });
+  const dirs = await fs.readdir(__dirname);
+  for (const dir of dirs) {
+    // Skip anything that doesn't match the expected directory name format.
+    const match = DIRECTORY_REGEX.exec(dir);
+    if (!match) continue;
 
-  // Check for duplicate UUIDs
-  _(news_items)
-    .groupBy('uuid')
-    .each(function (aList, uuid) {
-      if (aList.length > 1) {
-        const directories = aList.map((a) => a.directory).join(', ');
-        throw new Error(`UUID ${uuid} is used in multiple news items: ${directories}`);
-      }
+    const info = await jsonLoad.readInfoJSON(
+      path.join(__dirname, dir, 'info.json'),
+      schemas.infoNewsItem,
+    );
+
+    news_items.push({
+      ...info,
+      directory: dir,
+      index: match[1],
     });
+  }
 
   // Check for duplicate UUIDs
-  _(news_items)
-    .groupBy('index')
-    .each(function (aList, index) {
-      if (aList.length > 1) {
-        const directories = aList.map((a) => a.directory).join(', ');
-        throw new Error(`News item index ${index} is used in multiple news items: ${directories}`);
-      }
-    });
+  Object.entries(_.groupBy(news_items, 'uuid')).forEach(([uuid, items]) => {
+    if (items.length > 1) {
+      const directories = items.map((a) => a.directory).join(', ');
+      throw new Error(`UUID ${uuid} is used in multiple news items: ${directories}`);
+    }
+  });
+
+  // Check for duplicate indexes
+  Object.entries(_.groupBy(news_items, 'index')).forEach(([index, items]) => {
+    if (items.length > 1) {
+      const directories = items.map((a) => a.directory).join(', ');
+      throw new Error(`News item index ${index} is used in multiple news items: ${directories}`);
+    }
+  });
 
   return _.sortBy(news_items, 'directory');
 }
 
-module.exports.initAsync = async function (notify_with_new_server) {
-  const lockName = 'news_items';
-  const lock = await namedLocks.waitLockAsync(lockName, {});
```

Comment:
While implementing this, I noticed something that's a little suspect: `waitLockAsync` does not appear to actually wait for the lock. The only time we don't use `SKIP LOCKED` is when a timeout is set, but in this case, the timeout is zero.

https://github.com/PrairieLearn/PrairieLearn/blob/8c1b71e9c7e187bafeefa4186e14f659f738a5e3/packages/named-locks/src/index.ts#L236-L238

This feels wrong to me, and also doesn't match how we've documented this functionality:

https://github.com/PrairieLearn/PrairieLearn/blob/8c1b71e9c7e187bafeefa4186e14f659f738a5e3/packages/named-locks/src/index.ts#L20-L21

I think we need to make one of two choices here (there may also be other options I haven't thought of):

- Decide that we like the current behavior and update the documentation to reflect that. This may be a reasonable choice; maybe we don't ever want locks to wait forever?
- Change the semantics of `timeout` as follows:
  - `timeout > 0`: don't use `SKIP LOCKED`, wait for `timeout` to acquire the lock
  - `timeout == 0`: use `SKIP LOCKED`, no waiting
  - `timeout == -1`: don't use `SKIP LOCKED`, disable the lock timeout entirely
 
On top of that, to bikeshed: all of this got me thinking that we might not even want `tryWithLock` to exist anymore, since it's just a very narrow wrapper around `doWithLock` that just sets `onNotAcquired` to return `null` if the lock couldn't be acquired, and in all but one usage of `tryWithLock` we already set `onNotAcquired` to something else.

Reply:
I tend to think that we never want to wait forever to acquire a lock and instead we should always handle failure in some way in application code (even if just by throwing an uncaught exception and crashing). That is, I think the code is working as we want and it's just the comment describing `timeout` that needs to be adjusted.

Regarding the bike-shedding, I agree that we don't want `tryWithLock` (it's confusingly similar to `doWithLock`).

I actually think we should bike-shed even harder and eliminate `tryLock`/`tryLockAsync` and `waitLock`/`waitLockAsync`. These made sense before we had `doWithLock`, but at this point I think we should always call `doWithLock`. If we really need extra control then we should call `getLock` directly, and in any case `tryLock`/`waitLock` are very thin wrappers around `getLock`. If we do eliminate `tryLock`/`waitLock` then we should ensure that `doWithLock` still defaults to `timeout: 0`, of course.

Reply:
> Regarding the bike-shedding, I agree that we don't want `tryWithLock` (it's confusingly similar to `doWithLock`).

This was actually a point of confusion for me in the recent past, as `tryWithLock` (`doWithLock` with a default of do nothing on not acquired) has nothing to do with `tryLock` (`getLock` with a timeout of 0), and there is no relationship between them.

I agree with eliminating tryLock/waitLock. I just did a quick check, and `tryLock` is only used (outside of the named-locks package itself) in the cron process, and `waitLock` is only used in news items, so eliminating should be very straightforward. `tryWithLock` is used in a few places but should also be easy enough to replace.

Reply:
I opened #9479 to do this. It'll have to wait until after this PR, which removed the final usage of `waitLockAsync`.